cluster-autoscaler/provisioningrequest/besteffortatomic/provisioning_class.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package besteffortatomic
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	"k8s.io/autoscaler/cluster-autoscaler/apis/provisioningrequest/autoscaling.x-k8s.io/v1beta1"
+	"k8s.io/autoscaler/cluster-autoscaler/clusterstate"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/core/scaleup"
+	"k8s.io/autoscaler/cluster-autoscaler/core/scaleup/orchestrator"
+	"k8s.io/autoscaler/cluster-autoscaler/estimator"
+	"k8s.io/autoscaler/cluster-autoscaler/processors/status"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/conditions"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/provreqclient"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+
+	ca_processors "k8s.io/autoscaler/cluster-autoscaler/processors"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// Best effort atomic provisionig class requests scale-up only if it's possible
+// to atomically request enough resources for all pods specified in a
+// ProvisioningRequest. It's "best effort" as it admits workload immediately
+// after successful request, without waiting to verify that resources started.
```

Comment:
I don't think the comment is correct. It's best effort because it's rely on AtomicIncreaseSize API and if it's not implemented it fallbacks to IncreaseSize. And for IncreaseSize API your comment is correct.

Reply:
It's still best effort unless `AtomicIncreaseSize` is supposed to wait for the resources to be provisioned (for the VMs to start and register as nodes) - a VM can still fail and get stuck.

If `AtomicIncreaseSize` is supposed to wait for everything to be ready, it would be a serious performance issue. We would have to wait for it asynchronously if we wanted to have a true guarantee.

Reply:
>If AtomicIncreaseSize is supposed to wait for everything to be ready

@kisieland I suppose it is, correct?

Reply:
`AtomicIncreaseSize` is not supposed to wait for everything to be ready, the API should be add X or do nothing. 
Whereas `IncreaseSize` is add X or anything that you are able to (anything in between 0 and X).

Reply:
What does "you are able to" mean in this context?

I have also been under the impression that `AtomicIncreaseSize` guarantees the VMs coming up reasonably soon if it succeeds. If not that, how is it different than regular `IncreaseSize`? At least in GCE, most issues with creating instances happen after the increase call (stockouts, quota issues, ip exhaustion, etc.). Is `AtomicIncreaseSize` supposed to account for them and guarantee they won't happen if the request succeeds?

Reply:
> If not that, how is it different than regular IncreaseSize? 

I believe the cloudprovider API within autoscaler doesn't specify that `IncreaseSize` has to do nothing but return an error if the entire delta is impossible to request. GCE cloudprovider happens to do that, though, so there's not much difference in this case.

The only way `AtomicIncreaseSize` could provide a stronger guarantee would be if it were a blocking call, and that's infeasible from performance perspective - if we want to wait for VMs before admitting workload, we'll need to find a way to do so without blocking other autoscaler ops. 

One way to do that would be to add a checkpoint to ProvisionionigRequest, sth like `TriggeredScaleUp=true`, then treat request as if it were of `check-capacity` class and only when the capacity is provisioned update it with `Provisioned=true`

Reply:
The main change of `AtomicIncreaseSize` vs the `IncreaseSize` is the fact that if the call would result in partial scale up (as it can happen with `IncreaseSize` for example due to stock-outs) then no VMs will be provisioned and the whole scale up will fail, i.e. no retries will be made.

Regarding the duration of the call, it should provide the answer within the similar time-frame as the  `IncreaseSize`. Which AFAIK now is also handled asynchronously.

Reply:
> The only way `AtomicIncreaseSize` could provide a stronger guarantee would be if it were a blocking call, and that's infeasible from performance perspective - if we want to wait for VMs before admitting workload, we'll need to find a way to do so without blocking other autoscaler ops.

I guess it could also e.g. provide a stronger guarantee asynchronously with a (much) longer timeout. I think that's what happens in the GKE ProvReq case right? Just want to understand the differences here.

> I believe the cloudprovider API within autoscaler doesn't specify that `IncreaseSize` has to do nothing but return an error if the entire delta is impossible to request. GCE cloudprovider happens to do that, though, so there's not much difference in this cas

@aleksandra-malinowska Hmmm I see, so the only difference is "atomic version guarantees that the method fails if it knows that not all VMs can be provisioned before doing the API call"? I have a couple of questions:

Do you really anticipate it being useful in practice, given that most issues impacting VM provisioning can't be detected before doing the increase call? The only thing I can see this catching is exceeding the node group's max limit, which is already covered by the scale-up code after your previous changes. Is this just for completeness with your previous changes?

> The main change of `AtomicIncreaseSize` vs the `IncreaseSize` is the fact that if the call would result in partial scale up (as it can happen with `IncreaseSize` for example due to stock-outs) then no VMs will be provisioned and the whole scale up will fail, i.e. no retries will be made.
> 
> Regarding the duration of the call, it should provide the answer within the similar time-frame as the `IncreaseSize`. Which AFAIK now is also handled asynchronously.

@kisieland This seems to go against what @aleksandra-malinowska is saying above? I haven't seen `IncreaseSize` calls failing because of stockouts - at least not in GCE which is the main target for this feature. In my experience, the increase call itself ~always succeeds unless the max MIG size limit is exceeded. Stockout issues are only visible when the VMs fail to come up after the call. Unless the atomic version is supposed to predict such stockout issues and fail the API call if so after all?

This isn't blocking for this PR, but IMO we should really clearly set the expectations here, both internally for the cloud providers and externally for ProvReq users. Even people deeply involved with the feature don't seem to have full clarity on this.

Reply:
The strength of guarantees for the atomicity depends on implementation in a given cloud provider. The call itself expresses only an intent. 

 `AtomicIncreaseSize` is used for scale-ups which only make sense if the entire delta is provisioned. 
 
 `IncreaseSize` is used when a partial scale-up might still be useful. This was always the default assumption in autoscaling logic, we're just making it explicit by introducing another option.

In GCE, we can use [ResizeRequest API](https://cloud.google.com/compute/docs/instance-groups/about-resize-requests-mig) for a more atomic resize. This still doesn't guarantee that the nodes will register with the KCP, but it helps avoid partial scale-ups in some scenarios, such as stockouts.

The assumption is that available external APIs are likely to differ substantially between providers, and so will the cases they handle or don't handle. Some providers may have no support for atomic resizes whatsoever, in which case the implementation will be identical to `IncreaseSize`'s. 

Because it's really only best effort on cloud provider's side, we don't want to promise that `AtomicIncreaseSize` gives any guarantee to autoscaling logic, and we don't rely on it later - failed scale-up is handled the exact same way.